qiskit/circuit/parameter.py

```diff
@@ -144,20 +140,27 @@ def __eq__(self, other):
         else:
             return False
 
+    def _hash_key(self):
+        # This isn't the entirety of the object that's passed to `hash`, just the "key" part of
+        # individual parameters.  The hash of a full `ParameterExpression` needs to depend on the
+        # "keys" of `Parameter`s, and our hash needs to be computable before we can be fully
+        # initialised as a `ParameterExpression`, so we break the cycle by making our "key"
+        # accessible separately.
+        return (self._name, self._uuid)
+
     def __hash__(self):
+        # This is precached for performance, since it's used a lot and we are immutable.
         return self._hash
 
+    # We have to manually control the pickling so that the hash is computable before the unpickling
```

Comment:
It took me a while to understand why the unpickling was putting this parameter into a hashmap. The issue is the `self._parameter_symbols = {self: symbol}` line in `__init__`. I am not sure how pickle avoids an infinite loop there when it pickles the attributes and slots of an object, but it must do something clever. When unpickling, it wants to recreate this dictionary but `self` is not fully initialized with a `_hash` when it does.

One alternative approach could be to override `_parameter_symbols` with a property that returns `{self: self._symbol_expr}` so that `self` is not in the data of the class instance. I think that would almost be enough to avoid `__getstate__` and `__setstate__`. However, I think we might still want to override `__getstate__` to null out `_hash` and have `__hash__` recompute it if it is `None`. The reason is to avoid pickling in one Python process and unpickling in another with a different hash seed and not having the restored `_hash` match what would be calculated for a fresh `hash()` call.

Reply:
Self-referential objects or just reference cycles in general aren't all that uncommon in Python - they're handled by all the standard library serialisers etc.  Even printing works on standard objects:
```python
>>> a = a[0] = [[]]
>>> a
[[...]]
```
(where that makes `a` a list that contains itself).  I think a pretty large number of reference cycles can be handled in `pickle` serialisation just by it inserting references to the various objects before it initialises each of them, and keeps a memo list of things it's still got to restore state to to avoid looping.  In the cases where there's a particular partial order between an object's attributes, that's when you need `__setstate__`.

We could potentially make `_parameter_symbols` a hashmap, but as you say, we'd still need to recompute the cached `_hash` on pickle restoration, so I'm not certain there's a huge reason to add the property - if nothing else, adding it as a descriptor means it needs to be recalculated each time it's accessed, and there's an overhead just from having the indirection.

Reply:
Perhaps self-reference is not uncommon, but it has the potential to result in code that is hard to understand, and I think that has become the case here. Can we break the cycle by making Parameter no longer inherit from ParameterExpression?

Reply:
It's obviously subjective, but personally I thought that this all interacted without much trouble with Python; it was just broken before because I think people weren't aware that there were data-model invariants they were required to uphold.  I think it maybe feels more complex _because_ the previous implementation was broken; this PR includes a few additional comments to highlight the invariants.

It's a little unusual to have a concrete class depend on a concrete subclass of itself (but common to have an abstract class depend on a concrete subclass), but it wouldn't be super easy to walk that back now; there's a lot of extant code that depends on being able to retrieve the `.parameters` as a set from an expression.